v-next/hardhat/src/internal/builtin-plugins/gas-analytics/gas-analytics-manager.ts

```diff
@@ -0,0 +1,315 @@
+import type { GasAnalyticsManager, GasMeasurement } from "./types.js";
+import type { TableItem } from "@nomicfoundation/hardhat-utils/format";
+
+import crypto from "node:crypto";
+import path from "node:path";
+
+import { divider, formatTable } from "@nomicfoundation/hardhat-utils/format";
+import {
+  ensureDir,
+  getAllFilesMatching,
+  readJsonFile,
+  remove,
+  writeJsonFile,
+} from "@nomicfoundation/hardhat-utils/fs";
+import chalk from "chalk";
+import debug from "debug";
+
+const gasStatsLog = debug(
+  "hardhat:core:gas-analytics:gas-analytics-manager:gas-stats",
+);
+
+interface ContractGasStats {
+  deployment?: { gas: number; size: number };
+  functions: Map<
+    string, // function name or signature (if overloaded)
+    GasStats
+  >;
+}
+
+type GasStatsByContract = Map<string, ContractGasStats>;
+
+interface GasStats {
+  min: number;
+  max: number;
+  avg: number;
+  median: number;
+  calls: number;
+}
+
+type GasMeasurementsByContract = Map<string, ContractGasMeasurements>;
+
+interface ContractGasMeasurements {
+  deployment?: { gas: number; size: number };
+  functions: Map<
+    string, // functionSig
+    number[]
+  >;
+}
+
+export class GasAnalyticsManagerImplementation implements GasAnalyticsManager {
+  public gasMeasurements: GasMeasurement[] = [];
+  readonly #gasStatsPath: string;
+
+  constructor(gasStatsPath: string) {
+    this.#gasStatsPath = gasStatsPath;
+  }
+
+  public addGasMeasurement(gasMeasurement: GasMeasurement): void {
+    this.gasMeasurements.push(gasMeasurement);
+
+    gasStatsLog(
+      "Added gas measurement",
+      JSON.stringify(gasMeasurement, null, 2),
+    );
```

Comment:
I don't think this is going to be useful debug information: it's too granular, and it could mostly be obtained from disk if necessary. I would remove it.

Reply:
Done in https://github.com/NomicFoundation/hardhat/pull/7472/commits/5f69b05fdb81aa2f6157fe297a2db6db1a69d500